File: src/lib/astronomy/countdown.ts
import { differenceInSeconds } from 'date-fns'
import type { MoonData, SunTimes } from '@/types'

export interface NextEvent {
  type: 'golden-hour-am' | 'golden-hour-pm' | 'blue-hour-am' | 'blue-hour-pm' | 'sunrise' | 'sunset' | 'moonrise' | 'moonset'
  label: string
  time: Date
  secondsUntil: number
}

const EVENT_LABELS: Record<NextEvent['type'], string> = {
  'golden-hour-am': 'Morning golden hour',
  'golden-hour-pm': 'Golden hour',
  'blue-hour-am': 'Morning blue hour',
  'blue-hour-pm': 'Blue hour',
  sunrise: 'Sunrise',
  sunset: 'Sunset',
  moonrise: 'Moonrise',
  moonset: 'Moonset',
}

export function getNextEvent(
  now: Date,
  sunTimes: SunTimes,
  moonData: MoonData,
): NextEvent | null {
  const candidates: Array<{ type: NextEvent['type']; time: Date }> = [
    { type: 'blue-hour-am', time: sunTimes.blueHourMorningStart },
    { type: 'sunrise', time: sunTimes.sunrise },
    { type: 'golden-hour-am', time: sunTimes.goldenHourMorningStart },
    { type: 'golden-hour-pm', time: sunTimes.goldenHourStart },
    { type: 'sunset', time: sunTimes.sunset },
    { type: 'blue-hour-pm', time: sunTimes.blueHourEveningStart },
  ]

  if (moonData.moonrise) {
    candidates.push({ type: 'moonrise', time: moonData.moonrise })
  }
  if (moonData.moonset) {
    candidates.push({ type: 'moonset', time: moonData.moonset })
  }

  const future = candidates
    .filter((candidate) => candidate.time.getTime() > now.getTime())
    .sort((a, b) => a.time.getTime() - b.time.getTime())

  if (future.length === 0) return null

  const next = future[0]
  return {
    type: next.type,
    label: EVENT_LABELS[next.type],
    time: next.time,
    secondsUntil: differenceInSeconds(next.time, now),
  }
}

export function formatCountdown(secondsUntil: number): string {
  if (secondsUntil <= 0) return 'Now'

  const hours = Math.floor(secondsUntil / 3600)
  const minutes = Math.floor((secondsUntil % 3600) / 60)

  if (hours > 0) {
    return `${hours}h ${minutes}m`
  }
  if (minutes > 0) {
    return `${minutes}m`
  }
  return '<1m'
}

File: ios/PhotoPro/PhotoPro/Lib/Astronomy/Countdown.swift
import Foundation

enum CountdownEventType: String {
    case goldenHourAM = "golden-hour-am"
    case goldenHourPM = "golden-hour-pm"
    case blueHourAM = "blue-hour-am"
    case blueHourPM = "blue-hour-pm"
    case sunrise
    case sunset
    case moonrise
    case moonset
    case meteorShower = "meteor-shower"

    var label: String {
        switch self {
        case .goldenHourAM: return "Morning golden hour"
        case .goldenHourPM: return "Golden hour"
        case .blueHourAM: return "Morning blue hour"
        case .blueHourPM: return "Blue hour"
        case .sunrise: return "Sunrise"
        case .sunset: return "Sunset"
        case .moonrise: return "Moonrise"
        case .moonset: return "Moonset"
        case .meteorShower: return "Meteor shower"
        }
    }
}

struct CountdownEvent {
    var type: CountdownEventType
    var label: String
    var time: Date
    var secondsUntil: Int
}

enum Countdown {
    /// Find the next upcoming event from sun times, moon data, and optionally meteor showers.
    static func getNextEvent(
        now: Date,
        sunTimes: SunTimes,
        moonData: MoonData,
        latitude: Double? = nil
    ) -> CountdownEvent? {
        var candidates: [(type: CountdownEventType, time: Date, label: String?)] = [
            (.blueHourAM, sunTimes.blueHourMorningStart, nil),
            (.sunrise, sunTimes.sunrise, nil),
            (.goldenHourAM, sunTimes.goldenHourMorningStart, nil),
            (.goldenHourPM, sunTimes.goldenHourStart, nil),
            (.sunset, sunTimes.sunset, nil),
            (.blueHourPM, sunTimes.blueHourEveningStart, nil),
        ]

        if let moonrise = moonData.moonrise {
            candidates.append((.moonrise, moonrise, nil))
        }
        if let moonset = moonData.moonset {
            candidates.append((.moonset, moonset, nil))
        }

        if let lat = latitude {
            if let nextMeteor = MeteorCalculator.getNextMeteorShower(fromDate: now, userLatitude: lat) {
                if nextMeteor.zhr >= 50 {
                    let daysUntil = nextMeteor.peakDate.timeIntervalSince(now) / 86400
                    if daysUntil <= 30 && daysUntil > 0 {
                        candidates.append((.meteorShower, nextMeteor.peakDate, nextMeteor.name))
                    }
                }
            }
        }

        let future = candidates
            .filter { $0.time > now }
            .sorted { $0.time < $1.time }

        guard let next = future.first else { return nil }

        let secondsUntil = Int(next.time.timeIntervalSince(now))

        return CountdownEvent(
            type: next.type,
            label: next.label ?? next.type.label,
            time: next.time,
            secondsUntil: secondsUntil
        )
    }
}
